superflore/generators/bitbake/yocto_recipe.py

```diff
@@ -51,11 +54,30 @@ def __init__(self, name, distro, src_uri, tar_dir):
         self.archive_name = None
         self.license_md5 = None
         self.tar_dir = tar_dir
-        self.downloadArchive()
-        self.src_sha256 = hashlib.sha256(
-            open(self.getArchiveName(), 'rb').read()).hexdigest()
-        self.src_md5 = hashlib.md5(
-            open(self.getArchiveName(), 'rb').read()).hexdigest()
+        if md5_cache is not None:
+            if self.getArchiveName() not in md5_cache:
+                self.downloadArchive()
```

Comment:
Also you download the archive in both the if and else, but then don't download it in the sha256 since it's already there.

Also if you load the cache at the higher level you should write it out again at the higher level instead of each time you process a new package and just update the internal (md5)_cache object for each computation. And write the new cache values out on exit.

It might be worth making a dedicated class that loads holds, and computes the hashes conditionally. It can be parameterized on the hash function. And then you just pass it the md5_cache.get_hash(filename) and it will query if it's stored else compute it. If md5_cache is a context manager, you can pass it the cache filename and hash function. And it will load the file going in and dump the cache on the way out automatically.

Reply:
> Also if you load the cache at the higher level you should write it out again at the higher level instead of each time you process a new package and just update the internal (md5)_cache object for each computation. And write the new cache values out on exit.

That definitely sounds like a more pythonic way to do this. I'll get right on that.